core/dbt/task/clone.py

```diff
@@ -0,0 +1,185 @@
+        for node in self.manifest.nodes.values():
+            if node.unique_id not in selected_uids:
+                continue
+            if node.is_relational and not node.is_ephemeral:
+                relation = adapter.Relation.create_from(self.config, node)
+                result.add(relation.without_identifier())
+
+                # cache the 'other' schemas too!
+                if node.state_relation:  # type: ignore
+                    other_relation = adapter.Relation.create_from_node(
+                        self.config, node.state_relation  # type: ignore
+                    )
+                    result.add(other_relation.without_identifier())
+
+        return result
+
+    def before_run(self, adapter, selected_uids: AbstractSet[str]):
+        with adapter.connection_named("master"):
+            # unlike in other tasks, we want to add information from the --state manifest *before* caching!
+            self.defer_to_manifest(adapter, selected_uids)
+            # only create *our* schemas, but cache *other* schemas in addition
+            schemas_to_create = super().get_model_schemas(adapter, selected_uids)
+            self.create_schemas(adapter, schemas_to_create)
+            schemas_to_cache = self.get_model_schemas(adapter, selected_uids)
+            self.populate_adapter_cache(adapter, schemas_to_cache)
+
+    @property
+    def resource_types(self):
+        if not self.args.resource_types:
+            return NodeType.refable()
+
+        values = set(self.args.resource_types)
+
+        if "all" in values:
+            values.remove("all")
+            values.update(NodeType.refable())
+
+        values = [NodeType(val) for val in values if val in NodeType.refable()]
+
+        return list(values)
+
+    def get_node_selector(self) -> ResourceTypeSelector:
+        resource_types = self.resource_types
+
+        if self.manifest is None or self.graph is None:
+            raise DbtInternalError("manifest and graph must be set to get perform node selection")
+        return ResourceTypeSelector(
+            graph=self.graph,
+            manifest=self.manifest,
+            previous_state=self.previous_state,
+            resource_types=resource_types,
+        )
+
+    def get_runner_type(self, _):
+        return CloneRunner
+
+    def _get_deferred_manifest(self) -> Optional[WritableManifest]:
+        state = self.previous_state
+        if state is None:
+            raise DbtRuntimeError(
+                "--state is required for cloning relations from another environment"
+            )
+
+        if state.manifest is None:
+            raise DbtRuntimeError(f'Could not find manifest in --state path: "{self.args.state}"')
+        return state.manifest
+
+    # Note that this is different behavior from --defer with other commands, which *merge*
+    # selected nodes from this manifest + unselected nodes from the other manifest
+    def defer_to_manifest(self, adapter, selected_uids: AbstractSet[str]):
```

Comment:
This method has the same name, but different behavior, in that it calls `Manifest.add_from_artifact` instead of `Manifest.merge_from_artifact`. Could give it a different name for clarity